polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs

```diff
@@ -878,113 +1095,147 @@ impl FragmentChain {
+						candidate.output_head_data_hash,
+						candidate.parent_head_data_hash,
+					))
+				});
+
+			// Choose the best candidate.
+			let best_candidate =
+				possible_children.min_by(|(_, ref child1, _, _), (_, ref child2, _, _)| {
+					// Always pick a candidate pending availability as best.
+					if self.scope.get_pending_availability(child1).is_some() {
+						Ordering::Less
+					} else if self.scope.get_pending_availability(child2).is_some() {
+						Ordering::Greater
+					} else {
+						// Otherwise, use the fork selection rule.
+						fork_selection_rule(child1, child2)
 					}
+				});
 
-					let f = Fragment::new(
-						relay_parent.clone(),
-						constraints,
-						// It's cheap to clone because it's wrapped in an Arc
-						candidate.candidate.clone(),
-					);
-
-					match f {
-						Ok(f) => f,
-						Err(e) => {
-							gum::debug!(
-								target: LOG_TARGET,
-								err = ?e,
-								?relay_parent,
-								candidate_hash = ?candidate.candidate_hash,
-								"Failed to instantiate fragment",
-							);
-
-							break
-						},
-					}
-				};
+			if let Some((fragment, candidate_hash, output_head_data_hash, parent_head_data_hash)) =
+				best_candidate
+			{
+				// Remove the candidate from storage.
+				storage.remove_candidate(&candidate_hash);
 
 				// Update the cumulative constraint modifications.
 				cumulative_modifications.stack(fragment.constraint_modifications());
 				// Update the earliest rp
-				earliest_rp = relay_parent;
+				earliest_rp = fragment.relay_parent().clone();
 
 				let node = FragmentNode {
 					fragment,
-					candidate_hash: candidate.candidate_hash,
+					candidate_hash,
+					parent_head_data_hash,
+					output_head_data_hash,
 					cumulative_modifications: cumulative_modifications.clone(),
 				};
 
-				self.chain.push(node);
-				self.candidates.insert(candidate.candidate_hash);
-				// We've already checked for forks and cycles.
-				self.by_parent_head
-					.insert(candidate.parent_head_data_hash(), candidate.candidate_hash);
-				self.by_output_head
-					.insert(candidate.output_head_data_hash(), candidate.candidate_hash);
-				added_child = true;
-				// We can only add one child for a candidate. (it's a chain, not a tree)
-				break;
-			}
-
-			if !added_child {
+				// Add the candidate to the chain now.
+				self.best_chain.push(node);
+			} else {
 				break
 			}
 		}
 	}
+
+	/// Mark a candidate as backed. Return `None` if the candidate is not part of the unconnected
+	/// storage.
+	/// This will trigger a recreation of the best backable chain.
+	pub fn candidate_backed(&self, newly_backed_candidate: &CandidateHash) -> Option<Self> {
+		// Get the storage containing both the backable chain and the unconnected candidates.
+		let mut old_storage = self.as_candidate_storage();
+
+		if !old_storage.mark_backed(newly_backed_candidate) {
+			return None
+		}
+
+		// Repopulate.
+		Some(Self::populate(self.scope.clone(), old_storage))
```

Comment:
While elegant, also quite inefficient. We do full rebuilding and pruning of store (breath first search) on each and every backed candidate. With many cores (NOMT, other PoV bound usecases) this adds up and I think it is also not necessary.

Instead of doing `Self::populate()`, we should do:

```rust
self.revert_chain_to(head_data: Hash); // newly introduced method traversing the chain backwards, dropping all elements (moving back to storage) until candidate with given head data is found.
self.populate_chain();
```

This way:

1. We don't do costly pruning on each and every backing.
2. Most likely case that the chain is built from start to finish, `revert_chain_to` will do nothing and `populate_chain` has very little work.
3. Unlikely case that we build in reverse order, we are still kind of efficient as there is at least no chain to traverse and we only got once through the storage to find a potential child.
4. In the more realistic case that candidates simply come slightly out of order, we get a mixture of (2) and (3), which should also be quite efficient.

Reply:
An initial prototype of mine did something similar. However, I decided that the added complexity was not worth it and I tried to keep it simple. Fair point, I'll reconsider this

Reply:
Done, I optimised this. Have a look!